agent/resources/applicationContext.xml

```diff
@@ -17,106 +17,89 @@
-          p:targetObject-ref="agentController"
-          p:targetMethod="loop"/>
+      </bean>
+    </property>
+  </bean>
 
-    <bean id="agentControllerPinger" class="org.springframework.scheduling.timer.MethodInvokingTimerTaskFactoryBean"
-          p:targetObject-ref="agentController"
-          p:targetMethod="ping"/>
+  <bean id="urlService" class="com.thoughtworks.go.util.URLService"/>
+  <bean id="httpService" class="com.thoughtworks.go.util.HttpService"/>
+  <bean id="artifactsManipulator" class="com.thoughtworks.go.publishers.GoArtifactsManipulator"/>
+  <bean id="sslInfrastructureService" class="com.thoughtworks.go.agent.service.SslInfrastructureService"/>
+  <bean id="agentUpgradeService" class="com.thoughtworks.go.agent.service.AgentUpgradeService"/>
+  <bean id="agentRegistry" class="com.thoughtworks.go.config.DefaultAgentRegistry"/>
+  <bean id="subprocessLogger" class="com.thoughtworks.go.util.SubprocessLogger"/>
+  <bean id="agentPluginsInitializer" class="com.thoughtworks.go.agent.AgentPluginsInitializer"/>
+  <bean id="packageAsRepositoryExtension"
+        class="com.thoughtworks.go.plugin.access.packagematerial.PackageAsRepositoryExtension"/>
+  <bean id="scmExtension" class="com.thoughtworks.go.plugin.access.scm.SCMExtension"/>
+  <bean id="taskExtension" class="com.thoughtworks.go.plugin.access.pluggabletask.TaskExtension"/>
+  <bean id="httpClient" class="com.thoughtworks.go.agent.common.ssl.GoAgentServerHttpClient">
+    <constructor-arg ref="httpClientBuilder"/>
+  </bean>
+  <bean id="httpClientBuilder" class="com.thoughtworks.go.agent.common.ssl.GoAgentServerHttpClientBuilder">
+    <constructor-arg ref="systemEnvironment"/>
+  </bean>
 
-    <bean id="agentControllerInstructionExecutor"
-          class="org.springframework.scheduling.timer.MethodInvokingTimerTaskFactoryBean"
-          p:targetObject-ref="agentController"
-          p:targetMethod="executeAgentInstruction"/>
+  <bean id="webSocketClientBuilder" class="com.thoughtworks.go.agent.common.ssl.GoAgentServerWebSocketClientBuilder">
+    <constructor-arg ref="systemEnvironment"/>
+  </bean>
 
+  <bean id="webSocketClientHandler" class="com.thoughtworks.go.agent.WebSocketClientHandler"/>
+  <bean id="webSocketSessionHandler" class="com.thoughtworks.go.agent.WebSocketSessionHandler"/>
 
-    <!-- worker thread -->
-    <bean name="loopThread" class="org.springframework.scheduling.timer.TimerFactoryBean">
-        <property name="scheduledTimerTasks">
-            <list>
-                <bean class="org.springframework.scheduling.timer.ScheduledTimerTask"
-                      p:delay="${agent.get.work.delay}"
-                      p:period="${agent.get.work.interval}"
-                      p:timerTask-ref="agentControllerLooper"/>
-            </list>
-        </property>
-    </bean>
+  <bean id="agentController" class="com.thoughtworks.go.agent.AgentController" factory-method="createInstance">
```

Comment:
You could create a `AgentControllerFactory` with non-static factory-method, then you wouldn't have to pass on `constructor-arg` from the xml file. You could just rely on spring to autowire the constructor arguments, and then a parameter-less factory method to instantiate the desired controller. The xml would look something like this then:
``` xml
  <bean id="agentControllerFactory" class="com.thoughtworks.go.agent.AgentControllerFactory" />
  <bean id="agentController" class="com.thoughtworks.go.agent.AgentController" factory-bean="agentControllerFactory" factory-method="createInstance" />
```

AgentControllerFactory.java:
``` java
@Component
public class AgentControllerFactory {
    private final BuildRepositoryRemote server;
    private final GoArtifactsManipulator manipulator;
    private final PluginManager pluginManager;
    private final SslInfrastructureService sslInfrastructureService;
    private final SystemEnvironment systemEnvironment;
    private final AgentRegistry agentRegistry;
    private final SubprocessLogger subprocessLogger;
    private final AgentUpgradeService agentUpgradeService;
    private final PackageAsRepositoryExtension packageAsRepositoryExtension;
    private final SCMExtension scmExtension;
    private final TaskExtension taskExtension;
    private final HttpService httpService;
    private final WebSocketClientHandler webSocketClientHandler;
    private final WebSocketSessionHandler sessionHandler;

    @Autowired
    public AgentControllerFactory(
            BuildRepositoryRemote server,
            GoArtifactsManipulator manipulator,
            AgentUpgradeService agentUpgradeService,
            PluginManager pluginManager,
            SslInfrastructureService sslInfrastructureService,
            SystemEnvironment systemEnvironment,
            AgentRegistry agentRegistry,
            SubprocessLogger subprocessLogger,
            PackageAsRepositoryExtension packageAsRepositoryExtension,
            SCMExtension scmExtension,
            TaskExtension taskExtension,
            HttpService httpService,
            WebSocketClientHandler webSocketClientHandler, WebSocketSessionHandler sessionHandler) {
        this.server = server;
        this.manipulator = manipulator;
        this.pluginManager = pluginManager;
        this.sslInfrastructureService = sslInfrastructureService;
        this.systemEnvironment = systemEnvironment;
        this.agentRegistry = agentRegistry;
        this.subprocessLogger = subprocessLogger;
        this.agentUpgradeService = agentUpgradeService;
        this.packageAsRepositoryExtension = packageAsRepositoryExtension;
        this.scmExtension = scmExtension;
        this.taskExtension = taskExtension;
        this.httpService = httpService;
        this.webSocketClientHandler = webSocketClientHandler;
        this.sessionHandler = sessionHandler;
        PluginManagerReference.reference().setPluginManager(pluginManager);
    }

    public AgentController createInstance() {
        if (systemEnvironment.isWebsocketEnabled()) {
            return new AgentWebSocketClientController(
                    server,
                    manipulator,
                    sslInfrastructureService,
                    agentRegistry,
                    agentUpgradeService,
                    subprocessLogger,
                    systemEnvironment,
                    pluginManager,
                    packageAsRepositoryExtension,
                    scmExtension,
                    taskExtension,
                    httpService, webSocketClientHandler, sessionHandler);
        } else {
            return new AgentHTTPClientController(
                    server,
                    manipulator,
                    sslInfrastructureService,
                    agentRegistry,
                    agentUpgradeService,
                    subprocessLogger,
                    systemEnvironment,
                    pluginManager,
                    packageAsRepositoryExtension,
                    scmExtension,
                    taskExtension);
        }
    }
}
```

Technically, it doesn't make much difference, but it keeps us from updating the xml file everytime the signature changes.